### PR TITLE
9 add authentication

### LINKED ISSUE #9: add authentication
- [x] add a `AETOS_ENABLE_IP_ALLOWLIST` django settings option
- [x] add a `AETOS_IP_ALLOWLIST` django settings option, `["127.0.0.1", "51.66.11.92"]`
- [x] add a `AETOS_ENABLE_AUTH` django settings option
- [x] add a `AETOS_AUTH_TOKENLIST` django settings option, `["igiu5losh3Opano", "yaeyueFaumeYoo8"]`
- [x] if `..._USE_IP_ALLOWLIST` is True, but there are no allowed IPs, always deny
- [x] if `..._USE_AUTH` is True, but there are no tokens, always deny
- [x] otherwise, only allow if all enabled auth methods pass

### PR DESCRIPTION
**Pull request by berthaottokarl**
Closes #9

* Authentication with Bearer token instead of Basic Auth

Additionally to the task list there's also: 
* Settings documentation in `README.rst`
* e2e tests in `example_project` for different auth settings configurations

If the test cases are enough - please pull.

### CODE REVIEW

File: example_project/example_project/settings.py [hunk 1]
```diff
@@ -124,3 +124,5 @@
 # https://docs.djangoproject.com/en/5.0/ref/settings/#default-auto-field
 
 DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
+
+# WARNING: Do not set AETOS_* settings, otherwise tests fail.
```
Comment (luto) [inline example_project/example_project/settings.py]: Line endings are inconsistent in this PR, did you activate the pre-commit hooks?

File: src/django_aetos/views.py [hunk 2]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
...
+    try:
+        authorization_header = request.META['HTTP_AUTHORIZATION']
+    except KeyError:
+        authorization_header = ""
+    type, sep, token = authorization_header.partition(' ')
+    allowed_tokens = app_settings.AUTH_TOKENLIST
+    
+    if app_settings.ENABLE_AUTH and type == "Bearer" and token in allowed_tokens:
+        return True
+    elif not app_settings.ENABLE_AUTH:
```
Comment (luto) [inline src/django_aetos/views.py]: probably use `is False` here so we don't interpret `None` (so no value at all) as false.

File: src/django_aetos/views.py [hunk 3]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
...
+    
+    return metrics
+
+def check_auth(request):
+    ip_address = request.META['REMOTE_ADDR']
+    allowed_ips = app_settings.IP_ALLOWLIST
+    
+    if app_settings.ENABLE_IP_ALLOWLIST and ip_address in allowed_ips:
+        return True
+    elif not app_settings.ENABLE_IP_ALLOWLIST:
```
Comment (luto) [inline src/django_aetos/views.py]: probably use `is False` here so we don't interpret `None` (so no value at all) as false.

File: src/django_aetos/views.py [hunk 4]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
...
             known_metrics.add(metric["name_without_labels"])
             metrics.append(metric)
+    
+    return metrics
+
+def check_auth(request):
+    ip_address = request.META['REMOTE_ADDR']
+    allowed_ips = app_settings.IP_ALLOWLIST
+    
+    if app_settings.ENABLE_IP_ALLOWLIST and ip_address in allowed_ips:
```
Comment (luto) [inline src/django_aetos/views.py]: probably use `is True` here so we don't interpret true-ish values from wrong configs as true.

File: src/django_aetos/views.py [hunk 5]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
...
-    return response
+def check_ips(request):
+    try:
+        authorization_header = request.META['HTTP_AUTHORIZATION']
+    except KeyError:
+        authorization_header = ""
+    type, sep, token = authorization_header.partition(' ')
+    allowed_tokens = app_settings.AUTH_TOKENLIST
+    
+    if app_settings.ENABLE_AUTH and type == "Bearer" and token in allowed_tokens:
```
Comment (luto) [inline src/django_aetos/views.py]: probably use `is True` here so we don't interpret true-ish values from wrong configs as true.

File: src/django_aetos/views.py [hunk 6]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
...
+            request,
+            "metrics/export.txt",
+            context={"metrics": collect_response()},
+            content_type="text/plain",
+        )
+        response.content = re.sub(b"\n+", b"\n", response.content)
+        
+        return response
+    else:
+        return HttpResponse(status=401)
```
Comment (luto) [inline src/django_aetos/views.py]: Add a response here. At least "token or IP check failed". If we wan't to be nice, log which one failed.

File: src/django_aetos/views.py [hunk 7]
```diff
@@ -18,12 +19,45 @@ def export_metrics(request):
                 metric["add_help"] = True
             known_metrics.add(metric["name_without_labels"])
             metrics.append(metric)
+    
+    return metrics
+
+def check_auth(request):
+    ip_address = request.META['REMOTE_ADDR']
```
Comment (luto) [inline src/django_aetos/views.py]: `REMOTE_ADDR` will not be correct, if django is behind a reverse proxy (so 99% of the time). Document somewhere that an additional app like [`django-xff`](https://pypi.org/project/django-xff/) might be needed.

**Review by luto — CHANGES_REQUESTED** [review]
Some of the tests seem to fail, have a look. I also added some comments.

File: src/django_aetos/views.py (hunk 7 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Added Info in 54dd922650a9dd91e7590ea2d97fd2fe899ee9ea

File: src/django_aetos/views.py (hunk 6 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Differentiated response about which check failed 1ddd9c87f8bb9c78f7dc3998e406cf8b562178f4

File: src/django_aetos/views.py (hunk 5 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Fixed in 1ddd9c87f8bb9c78f7dc3998e406cf8b562178f4

File: src/django_aetos/views.py (hunk 4 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Fixed in 1ddd9c87f8bb9c78f7dc3998e406cf8b562178f4

File: src/django_aetos/views.py (hunk 3 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Fixed in 1ddd9c87f8bb9c78f7dc3998e406cf8b562178f4

File: src/django_aetos/views.py (hunk 2 above)
Comment (berthaottokarl) [reply src/django_aetos/views.py]: Fixed in 1ddd9c87f8bb9c78f7dc3998e406cf8b562178f4

File: example_project/example_project/settings.py (hunk 1 above)
Comment (berthaottokarl) [reply example_project/example_project/settings.py]: Activated pre-commit hooks now 1ca452cc1ac1a01b440639f76510e6c3e4813c12

**Pull request merged**